python.d/linux_power_supply.chart.py

```diff
@@ -0,0 +1,159 @@
+    # attributes.
+    return _get_generic_chart(syspath, 'energy', 'µWh', 'full', 'empty')
+
+def _get_voltage_chart(syspath):
+    # Voltage is measured in microvolts. We track up to five attributes.
+    return _get_generic_chart(syspath, 'voltage', 'µV', 'min', 'max')
+
+# This is a list of functions for generating charts.  Used below to save
+# a bit of code (and to make it a bit easier to add new charts).
+GET_CHART = {
+    'capacity': _get_capacity_chart,
+    'charge': _get_charge_chart,
+    'energy': _get_energy_chart,
+    'voltage': _get_voltage_chart
+}
+
+# This opens the specified file and returns the value in it or None if
+# the file doesn't exist.
+def _get_sysfs_value(filepath):
+    try:
+        with open(filepath, 'r') as datasource:
+            return int(datasource.read())
+    except (OSError, IOError):
+        return None
+
+# Certain attributes (*_empty and *_empty_design) are implicitly zero
+# if not provided by the driver.  This function is used to fetch those
+# attributes.
+def _get_sysfs_value_or_zero(filepath):
+    result = _get_sysfs_value(filepath)
+    if result is None:
+        return 0
+    return result
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.definitions = dict()
+        self.order = list()
+        self.attrlist = list()
+        self.supply = self.configuration.get('supply', None)
+        if self.supply is not None:
+            self.syspath = '/sys/class/power_supply/{0}'.format(self.supply)
+        self.types = self.configuration.get('charts', 'energy charge capacity voltage').split()
+
+    def check(self):
+        if self.supply is None:
+            self.error('No power supply specified for monitoring.')
+            return False
+        if not len(self.types) > 0:
+            self.error('No attributes requested for monitoring.')
+            return False
+        if os.access(self.syspath, os.R_OK):
+            chartset = set(GET_CHART.keys()).intersection(set(self.types))
+            if len(chartset) > 0:
+                charts = dict()
+                attrlist = list()
+                for item in chartset:
+                    chart, attrs = GET_CHART[item](self.syspath)
+                    if chart is not None:
+                        charts.update(chart)
+                        attrlist.extend(attrs)
+                if len(charts) == 0:
+                    self.error('No charts can be created.')
+                    return False
+                self.definitions.update(charts)
+                self.order.extend(sorted(charts.keys()))
+                self.attrlist.extend(attrlist)
+            else:
+                self.error('No valid attributes requested for monitoring.')
+                return False
+        else:
+            self.error('Unable to access {0}'.format(self.syspath))
+            return False
+        return True
```

Comment:
I suggest to rewrite `check()`
Nested `if else` is always bad. Please avoid it.
```python
if not if os.access(self.syspath, os.R_OK):
    self.error('No valid attributes requested for monitoring.')
    return None

charset = set(GET_CHART.keys()).intersection(set(self.types))
if not charset:
    self.error('No valid attributes requested for monitoring.')
    return None
...
...
...
```
Keep in mind - empty list is always `False`, non empty is always `True`. Do not use `len`.